### PR TITLE
Fix fatal error in WP_Fonts_Resolver::get_settings()

### PR DESCRIPTION
**Pull request by anton-vlasenko**
## What?
This PR aims to fix a PHP fatal error that occurs within `WP_Fonts_Resolver::get_settings()` under certain conditions.
```
PHP Fatal error:  Uncaught TypeError: array_merge(): Argument #1 must be of type array, null given in /lib/experimental/fonts-api/class-wp-fonts-resolver.php:208
```

## Why?
Generally, the code should not trigger PHP fatal errors because certain variables do not match the expected type.

## How?
This PR ensures that the `array_merge()` function inside `WP_Fonts_Resolver::get_settings()` always receives arrays as input parameters.

## Testing instructions.
1. Activate a `Pendant` block theme.
2. Add
```php
define('FONT_LIBRARY_DISABLED', true);
```
to your config file.
3. Edit the src/wp-content/themes/pendant/styles/dark-navy.json file by replacing the following JSON block:
```json
"typography": {
			"fontFamilies": [
				{
					"fontFamily": "Inter, sans-serif",
					"slug": "body-font",
					"name": "Body",
					"fontFace": [
						{
							"fontDisplay": "block",
							"fontFamily": "Inter",
							"fontWeight": "100 500",
							"fontStyle": "normal",
							"fontStretch": "normal",
							"src": [
								"file:./assets/fonts/Inter-Light.ttf"
							]
						},
						{
							"fontDisplay": "block",
							"fontFamily": "Inter",
							"fontWeight": "600 900",
							"fontStyle": "normal",
							"fontStretch": "normal",
							"src": [
								"file:./assets/fonts/Inter-Bold.ttf"
							]
						}
					]
				},
				{
					"fontFamily": "'InterLight', serif",
					"slug": "heading-font",
					"name": "Headings",
					"fontFace": [
						{
							"fontDisplay": "block",
							"fontFamily": "InterLight",
							"fontStyle": "normal",
							"fontStretch": "normal",
							"src": [
								"file:./assets/fonts/Inter-Light.ttf"
							]
						}
					]
				}
			]
		}
	},
```
with
```json
"typography": {
			"fontFamilies": 1
		}
```
4. Make sure you see no
```
Warning: array_merge(): Expected parameter 2 to be an array,
```
error in the admin panel.

### CODE REVIEW

**Comment by simison** [comment]
Alternative https://github.com/WordPress/gutenberg/pull/55974

File: lib/experimental/fonts-api/class-wp-fonts-resolver.php
Comment (azaozz) [inline lib/experimental/fonts-api/class-wp-fonts-resolver.php]: Thinking this can be outside (above) the loop as `$settings` is not changed in the loop, just a bit of optimization.

At the same time not sure what is the expected default value (and value type) of `$settings['typography']['fontFamilies']['theme']` when it is empty or not set. Same for `$variation['settings']['typography']['fontFamilies']['theme']`. Setting these to empty arrays when missing means that the final `$settings['typography']['fontFamilies']['theme']` may be an empty array. Is that expected?

**Comment by azaozz** [comment]
> Alternative https://github.com/WordPress/gutenberg/pull/55974

Yep, was going to comment there that it may make sense to check both values separately. This PR does that. 

The other thing I'm not sure about is what value(s) and type(s) are expected at `$settings['typography']['fontFamilies']['theme']`, and/or can it be missing/unset.

File: lib/experimental/fonts-api/class-wp-fonts-resolver.php
Comment (anton-vlasenko) [reply lib/experimental/fonts-api/class-wp-fonts-resolver.php]: Good catch! Fixed in 8e94c1c6f6665546815beddabf21b847ce3648a3 and 061bafb4427e51bbcaebf847beb0017eaa914dda.

**Review by mikachan — COMMENTED** [review]
The changes here look good to me and I believe they'll prevent the fatal error in question. I'm happy to approve if needed, although this isn't my area of expertise.

**Comment by azaozz** [comment]
Looking  bit more at `get_settings()`, it's starting to look pretty buggy/inconsistent. Wondering if now would be a good time to fix it or just patch the obvious bug and leave the rest for later.

This is at the top of the loop:
```
if ( empty( $variation['settings']['typography']['fontFamilies'] ) ) {
	continue;
}
```
Not sure why it only checks `$variation['settings']['typography']['fontFamilies']` when we actually expect to use/merge`$variation['settings']['typography']['fontFamilies']['theme']`? Doesn't make sense imho. Should probably be checking if `theme` is empty. That will also simplify the code after it.

This is at the bottom of the loop:
```
// Make sure there are no duplicates.
$settings['typography']['fontFamilies'] = array_unique( $settings['typography']['fontFamilies'] );
```
What is it supposed to be checking for duplicates? We've likely added some array elements to `$settings['typography']['fontFamilies']['theme']` so the above code does nothing at all. Then why would it do that inside the loop, and not after it when the array has been set/merged with all the variations?

**Comment by anton-vlasenko** [comment]
> What is it supposed to be checking for duplicates? We've likely added some array elements to $settings['typography']['fontFamilies']['theme'] so the above code does nothing at all. Then why would it do that inside the loop, and not after it when the array has been set/merged with all the variations?

@azaozz I noticed it too and agree with your concerns.
Since I'm not the author of the code, it's hard to determine the original intent.
As this PR primarily focuses on resolving the fatal errors in WP_Fonts_Resolver::get_settings(), perhaps it would be better to address these concerns in another PR.

**Review by vcanales — APPROVED** [review]
> As this PR primarily focuses on resolving the fatal errors in WP_Fonts_Resolver::get_settings(), perhaps it would be better to address these concerns in another PR.

+1 to this, I think a refactor should be addressed after getting the fatal errors being thrown out of the way. 

I've [created an issue](https://github.com/WordPress/gutenberg/issues/56044) so that we don't lose track of the discussion.

**Comment by noahtallen** [comment]
I hope I didn't step on any toes here -- this issue has been causing quite a few problems, so getting it into a patch release feels urgent enough to warrant moving fast!

**Comment by anton-vlasenko** [comment]
Regarding backporting to Core: this PR should be backported as part of a larger PR that includes backporting the `/experimental/fonts-api` code to Core.
It doesn't make sense to backport this single PR to Core, as the `WP_Fonts_Resolver` class does not yet exist in Core yet.

**Comment by mikachan** [comment]
> Regarding backporting to Core: this PR should be backported as part of a larger PR that includes backporting the /experimental/fonts-api code to Core.

To confirm further, this file is actually about to be removed and so will not require backporting at all. Please see this PR for more info: https://github.com/WordPress/gutenberg/pull/57972

**Comment by getdave** [comment]
✅ I updated the [PHP Sync Tracking Issue](https://github.com/WordPress/gutenberg/issues/57959) to note this PR does not require a backport.

**Pull request merged**
